src/test/java/org/folio/entitlement/it/KeycloakRetriesIT.java

```diff
@@ -0,0 +1,212 @@
+        .content(asJsonString(entitlementRequest));
+    queryParams.forEach(request::queryParam);
+
+    stubPost(wireMockClient, 1, urlEqualTo("/admin/realms/test/clients/null/authz/resource-server/scope"), null, 201);
+    stubAnyHttpMethod(wireMockClient, 1,
+      urlMatching("/admin/realms/test/clients/null/authz/resource-server/resource.*"), null, 500);
+
+    mockMvc.perform(request).andExpect(status().isBadRequest()).andExpect(
+      jsonPath("$.errors[0].parameters[0].value").value(
+        containsString("Failed to update authorization resources in Keycloak")));
+
+    var endpointsCalled =
+      wireMockClient.getServeEvents().stream().filter(e -> e.getResponse().getStatus() == 500).toList();
+    assertThat(endpointsCalled.stream().filter(
+      e -> e.getRequest().getUrl().equals("/admin/realms/test/clients/null/authz/resource-server/resource")
+        && e.getRequest().getBodyAsString().contains("GET#folio-module1.events.item.get")).count()).isEqualTo(3);
+    assertThat(endpointsCalled.stream().filter(
+      e -> e.getRequest().getUrl().equals("/admin/realms/test/clients/null/authz/resource-server/resource")
+        && e.getRequest().getBodyAsString().contains("POST#folio-module1.events.item.post")).count()).isEqualTo(3);
+  }
+
+  @Test
+  @Sql("classpath:/sql/folio-entitlement.sql")
+  @WireMockStub(scripts = {"/wiremock/mgr-tenants/test/get.json",
+    "/wiremock/mgr-applications/folio-app2/get-by-ids-query-full.json",
+    "/wiremock/mgr-applications/folio-app2/get-discovery.json", "/wiremock/folio-module2/uninstall.json"})
+  void verifyModuleDeregistrationRetry() throws Exception {
+    var moduleId = "folio-module2-2.0.0";
+    var routeId = "mockroute1";
+
+    var wireMockClient = mockBaseDependencies(moduleId, routeId);
+    stubGet(wireMockClient, 1, urlMatching("/admin/realms/test/clients/null/authz/resource-server/resource.*"), null,
+      500);
+
+    var entitlementRequest = entitlementRequest(FOLIO_APP2_ID);
+    var queryParams = Map.of("purge", "true", "ignoreErrors", "true");
+    var request =
+      delete(updatePathWithPrefix("/entitlements")).contentType(APPLICATION_JSON).header(TOKEN, getSystemAccessToken())
+        .content(asJsonString(entitlementRequest));
+    queryParams.forEach(request::queryParam);
+
+    mockMvc.perform(request).andExpect(status().isBadRequest()).andExpect(
+      jsonPath("$.errors[0].parameters[0].value").value(
+        containsString("Failed to remove authorization resources in Keycloak")));
+
+    var endpointsCalled = wireMockClient.getServeEvents().stream().filter(e -> e.getResponse().getStatus() == 500)
+      .map(e -> e.getRequest().getUrl()).toList();
+    assertThat(endpointsCalled).hasSize(6);
+    assertThat(endpointsCalled.stream().filter(
+        v -> v.equals("/admin/realms/test/clients/null/authz/resource-server/resource?name=%2Ffolio-module2%2Fevents"))
```

Comment:
why `clientId` is null? use some predefined UUID to have keycloak-like request

also the stream can be extracted to have more readable code to the method like `filterCalledEndpointsByName(String name))`

Reply:
Filter is much more readable IMO, because every Java programmer knows it. Having some custom functionality instead will make it hard to understand what it does without checking it's source code.